Write data and vertices as floats

gassmoeller opened the pull request:
This is not yet ready, but I would like to ask for opinions before moving forward with this PR.

The output patches that are handed over to DataOut only contain data in single precision, but when writing the .vtu output format the data is implicitly converted to double precision, which seems like a waste of disk space. While the zlib compression reduces the overhead (the zero bits are compressed a lot) it is still 10% or so larger than necessary. More importantly: The position of vertices is written in double precision as well, which I feel is not necessary either. This can make up a lot of the total output written (imagine a 3D model with only one scalar field), so we might decide to always write this as single precision, or make it optional.

Writing the data as single precision should not be controversial (as it is previously converted to single precision anyway), but I would like to ask for opinions on the following points:
1. Should the vertex positions also always be single precision, or should this be optional (e.g. an option in the `VtkFlags`)?
2. Is there a reason that I do not see for first storing the data as float, but then writing as double?

For reference: The current state saves me about 30% of disk space for typical models (which can mean tens to hundreds of GB for my largest model series).

bangerth commented:
I had to do a double-take to recall that in the `Patch` data structure, we really do only store `float` objects. Given this, it clearly makes no sense to use `double` anywhere downstream.

I could come up with hypothetical reasons to use `double` for vertex locations, but since nobody seems to have ever complained about `float` for the data, I see no reason to believe that `float` isn't also good enough for vertex data. `float` in general has a range of only `1e-38` to `1e+38`. That's not large, but probably large enough for what most people want to do. So I'd say go ahead, and we can see whether anyone complains.

gassmoeller commented:
I updated the .pvtu and .pvd output functions, and updated the obvious tests (but I did not look through the whole test suite, I guess the tester will find out). If nobody objects the code is ready from my side.

jppelteret commented:
Thanks for taking the time to do this @gassmoeller! May I suggest that we typedef this in `base/types.h`, both in order to ensure consistency and just in case we (or a user) wish to have higher precision in the future?

jppelteret commented:
Now that I've written that, I see that it may cause an issue with this patch unless we link the data array type `Float32/Float64` to this typedef. If it seems like a pain to do (or just outright unnecessary) then lets ignore my suggestion.

masterleinad commented:
/run-tests

masterleinad commented:
At least the following test are failing:
```
hp/solution_transfer_03
hp/solution_transfer_05
grid/grid_out_08
grid/grid_out_07
```

gassmoeller commented:
I updated the test results. 
@jppelteret: I thought about that, but as you mentioned it is not a simple change. Alternatively, one could have a flag in the `VTKFlags` that determines the output precision. This would be simple to introduce into the `write_vtu_main` function, but it would not work for the `write_pvtu_record` function, which currently does not get the `VTKFlags` so this would require an interface change. Something for the future to-do list I guess.

jppelteret commented:
@gassmoeller Thanks for indicating what changes would be necessary to make this more flexible. I'm happy for us to move forward with your amendment and revisit this if its an issue for someone.

bangerth commented:
This looks good to me. Can you add a change log entry? Other than that, I think this is ready to merge!

gassmoeller commented:
Done.
